packages/fluentui/react-northstar/src/components/Carousel/Carousel.tsx

```diff
@@ -128,167 +129,174 @@ export const carouselSlotClassNames: CarouselSlotClassNames = {
-    ...commonPropTypes.createCommon({
-      content: false,
+export const Carousel: React.FC<WithAsProp<CarouselProps>> &
+  FluentComponentStaticProps<CarouselProps> & {
+    Item: typeof CarouselItem;
+    Navigation: typeof CarouselNavigation;
+    NavigationItem: typeof CarouselNavigationItem;
+    Paddle: typeof CarouselPaddle;
+  } = props => {
+  const context: ProviderContextPrepared = React.useContext(ThemeContext);
+  const { setStart, setEnd } = useTelemetry(Carousel.displayName, context.telemetry);
+  setStart();
+  const {
+    accessibility,
+    items,
+    circular,
+    getItemPositionText,
+    paddlePrevious,
+    paddleNext,
+    navigation,
+    thumbnails,
+    children,
+    ariaRoleDescription,
+    ariaLabel,
+    className,
+    design,
+    styles,
+    variables,
+    defaultActiveIndex,
+  } = props;
+
+  const ElementType = getElementType(props);
+
+  const { state, actions } = useStateManager<CarouselState, CarouselActions>(createCarouselManager, {
+    mapPropsToInitialState: () => ({
+      activeIndex: defaultActiveIndex,
     }),
-    activeIndex: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
-    ariaRoleDescription: PropTypes.string,
-    ariaLabel: PropTypes.string,
-    circular: PropTypes.bool,
-    defaultActiveIndex: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
-    getItemPositionText: PropTypes.func,
-    items: customPropTypes.collectionShorthand,
-    navigation: PropTypes.oneOfType([customPropTypes.collectionShorthand, customPropTypes.itemShorthand]),
-    navigationPosition: PropTypes.string,
-    onActiveIndexChange: PropTypes.func,
-    paddleNext: customPropTypes.itemShorthand,
-    paddlesPosition: PropTypes.string,
-    paddlePrevious: customPropTypes.itemShorthand,
-    thumbnails: PropTypes.bool,
-  };
-
-  static autoControlledProps = ['activeIndex'];
-
-  static defaultProps = {
-    accessibility: carouselBehavior as Accessibility,
-    paddlePrevious: {},
-    paddleNext: {},
-  };
+    mapPropsToState: () => ({
+      activeIndex: props.activeIndex,
+    }),
+  });
 
-  static Item = CarouselItem;
-  static Navigation = CarouselNavigation;
-  static NavigationItem = CarouselNavigationItem;
-  static Paddle = CarouselPaddle;
+  const { prevActiveIndex, ariaLiveOn, itemIds, shouldFocusContainer, isFromKeyboard, activeIndex } = state;
 
-  static getAutoControlledStateFromProps(props: CarouselProps, state: CarouselState) {
-    const { items } = props;
-    const { itemIds } = state;
+  const itemRefs = React.useMemo<React.RefObject<HTMLElement>[]>(
+    () => Array.from({ length: items?.length }, () => React.createRef()),
+    [items?.length],
+  );
 
-    if (!items) {
-      return null;
-    }
+  // @TODO: Move the id generation to CarouselItem
+  React.useEffect(() => {
+    actions.updateItemIds(
+      items?.map((item, index) => getOrGenerateIdFromShorthand('carousel-item-', item, state.itemIds[index])),
+    );
+  }, [items?.length]);
```

Comment:
This will produce double render, this is no way.

Reply:
Adding it to `mapPropsToState` tests break complaing that `state` is not defined ( we would be trying to access state before defining it )

Reply:
Removed `itemIds` at all after agreed with @kolaps33 and @silviuaavram